File: lecture/algorithms/remind/mergesort/01.js
function mergeSort(arr){
  if(arr.length<2) return arr
  let mid = Math.floor(arr.length/2)
  let firstArr = arr.slice(0,mid)
  let secondArr = arr.slice(mid)
  return merge(mergeSort(firstArr),mergeSort(secondArr))
}


function merge(arr1,arr2){
  let temp=0
  let answer = []
  while(arr1.length&&arr2.length){
    if(arr1[0]<arr2[0]) temp = arr1.shift()
    else temp=arr2.shift()
    answer.push(temp)
  }
  if(arr1.length+arr2.length!==0){
    answer=answer.concat(arr1.concat(arr2))
  }
  return answer
}

console.log(mergeSort([1,4,3,2,6,4,7,99]))

File: lecture/algorithms/remind/mergesort/01.py
def mergeSort(arr):
    mid = len(arr)/2
    firstArr = arr[:mid]
    secondArr = arr[mid:]
    if len(arr)<2: 
        return arr
    else:
        return merge(mergeSort(firstArr),mergeSort(secondArr))


def merge(arr1,arr2):
    answer=[]
    while(len(arr1) and len(arr2)):
        if arr1[0] < arr2[0]:
            temp = arr1.pop(0)
            answer.append(temp)
        else:
            answer.append(arr2.pop(0))
    if len(arr1) or len(arr2):
        arr1.extend(arr2)
        answer.extend(arr1)
    return answer

print(mergeSort([5,4,21,7,4,7,2,4,7,79]))
